File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/scan/PageEditorScreen.kt
/*
 * PageEditorScreen.kt
 *
 * Page-editing surface — re-crop, rotate, retake. Reachable from
 * `ScanReviewScreen` after a capture completes (before the user
 * taps Done) and from `NoteEditorScreen`'s page tab when the user
 * wants to revise an existing scan.
 *
 * Wireframe today — the actual image transform pipeline (Bitmap
 * rotation, draggable crop corners, page re-capture via
 * GmsDocumentScanning) lands in a follow-up. Shipping the picker
 * chrome now means the route + theme + control surface are
 * settled when transforms slot in.
 *
 * Mirror of iOS `PageEditorScreen.swift`.
 */

package app.quickink.mobile.features.scan

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Crop
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.RotateRight
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.quickink.mobile.ui.theme.LocalQuickInkColors
import app.quickink.mobile.ui.theme.LocalQuickInkTypography
import app.quickink.mobile.ui.theme.QuickInkRadius
import app.quickink.mobile.ui.theme.QuickInkSpacing

private enum class EditorTool { Crop, Rotate, Retake }

@Composable
fun PageEditorScreen(
    onCancel: () -> Unit,
    onSave: () -> Unit,
    onRetake: () -> Unit,
) {
    val colors = LocalQuickInkColors.current
    val type = LocalQuickInkTypography.current

    var rotationDeg by remember { mutableFloatStateOf(0f) }
    var activeTool by remember { mutableStateOf(EditorTool.Crop) }
    val animatedRotation by animateFloatAsState(targetValue = rotationDeg, label = "page-rotation")

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(colors.bg),
    ) {
        // Top bar
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = QuickInkSpacing.s2, vertical = QuickInkSpacing.s2),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                text     = "Cancel",
                style    = type.label,
                color    = colors.inkSoft,
                modifier = Modifier
                    .clickable(onClick = onCancel)
                    .padding(QuickInkSpacing.s3),
            )
            Spacer(Modifier.weight(1f))
            Text(text = "Edit page", style = type.label, color = colors.ink)
            Spacer(Modifier.weight(1f))
            Text(
                text     = "Save",
                style    = type.label,
                color    = colors.accent,
                modifier = Modifier
                    .clickable(onClick = onSave)
                    .padding(QuickInkSpacing.s3),
            )
        }

        // Page canvas
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f)
                .padding(QuickInkSpacing.s5),
            contentAlignment = Alignment.Center,
        ) {
            // Placeholder lined-paper page (real captured image
            // slots in here once the scan pipeline hands a Uri
            // through).
            var canvasSize by remember { mutableStateOf(androidx.compose.ui.geometry.Size.Zero) }
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .onGloballyPositioned { coords ->
                        canvasSize = androidx.compose.ui.geometry.Size(
                            coords.size.width.toFloat(),
                            coords.size.height.toFloat(),
                        )
                    }
                    .shadow(8.dp, RoundedCornerShape(QuickInkRadius.md))
                    .clip(RoundedCornerShape(QuickInkRadius.md))
                    .background(colors.surface)
                    .border(1.dp, colors.border, RoundedCornerShape(QuickInkRadius.md))
                    .rotate(animatedRotation),
                contentAlignment = Alignment.Center,
            ) {
                // Lined-paper rules + coral margin.
                androidx.compose.foundation.Canvas(modifier = Modifier.fillMaxSize().padding(16.dp)) {
                    val lineColor = colors.ink.copy(alpha = 0.10f)
                    val spacing = 16.dp.toPx()
                    var y = spacing
                    while (y < size.height) {
                        drawLine(
                            color       = lineColor,
                            start       = Offset(0f, y),
                            end         = Offset(size.width, y),
                            strokeWidth = 0.5f,
                        )
                        y += spacing
                    }
                }
                Box(
                    modifier = Modifier
                        .padding(start = 32.dp, top = 16.dp, bottom = 16.dp)
                        .width(1.5.dp)
                        .fillMaxSize()
                        .background(colors.accent.copy(alpha = 0.6f)),
                )
                Text(
                    text  = "scanned page",
                    style = type.handwritten.copy(fontSize = 28.sp),
                    color = colors.inkSoft.copy(alpha = 0.7f),
                )
            }

            // Crop handles overlay — only when crop tool active.
            if (activeTool == EditorTool.Crop) {
                CropHandle(modifier = Modifier.align(Alignment.TopStart).padding(16.dp))
                CropHandle(modifier = Modifier.align(Alignment.TopEnd).padding(16.dp))
                CropHandle(modifier = Modifier.align(Alignment.BottomStart).padding(16.dp))
                CropHandle(modifier = Modifier.align(Alignment.BottomEnd).padding(16.dp))
            }
        }

        // Tool strip (segmented)
        Row(
            modifier = Modifier
                .padding(horizontal = QuickInkSpacing.s5)
                .fillMaxWidth()
                .clip(RoundedCornerShape(QuickInkRadius.pill))
                .background(colors.borderSoft)
                .padding(4.dp),
            horizontalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            ToolPill(
                kind     = EditorTool.Crop,
                icon     = Icons.Filled.Crop,
                label    = "Crop",
                active   = activeTool == EditorTool.Crop,
                onClick  = { activeTool = EditorTool.Crop },
                modifier = Modifier.weight(1f),
            )
            ToolPill(
                kind     = EditorTool.Rotate,
                icon     = Icons.Filled.RotateRight,
                label    = "Rotate",
                active   = activeTool == EditorTool.Rotate,
                onClick  = {
                    activeTool = EditorTool.Rotate
                    rotationDeg = (rotationDeg + 90f) % 360f
                },
                modifier = Modifier.weight(1f),
            )
            ToolPill(
                kind     = EditorTool.Retake,
                icon     = Icons.Filled.Refresh,
                label    = "Retake",
                active   = activeTool == EditorTool.Retake,
                onClick  = {
                    activeTool = EditorTool.Retake
                    onRetake()
                },
                modifier = Modifier.weight(1f),
            )
        }
        Spacer(Modifier.size(QuickInkSpacing.s4))

        // Action hint row
        Row(
            modifier              = Modifier
                .fillMaxWidth()
                .padding(horizontal = QuickInkSpacing.s5)
                .padding(bottom = QuickInkSpacing.s7),
            verticalAlignment     = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.Center,
        ) {
            Icon(
                imageVector       = Icons.Filled.Info,
                contentDescription = null,
                tint              = colors.muted,
                modifier          = Modifier.size(12.dp),
            )
            Spacer(Modifier.size(QuickInkSpacing.s1))
            Text(
                text  = "Drag the corners to crop · Rotate steps 90°",
                style = type.caption,
                color = colors.muted,
            )
        }
    }
}

@Composable
private fun CropHandle(modifier: Modifier = Modifier) {
    val colors = LocalQuickInkColors.current
    Box(
        modifier = modifier
            .shadow(6.dp, CircleShape)
            .size(14.dp)
            .clip(CircleShape)
            .background(colors.accent)
    )
}

@Composable
private fun ToolPill(
    kind: EditorTool,
    icon: ImageVector,
    label: String,
    active: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = LocalQuickInkColors.current
    val type = LocalQuickInkTypography.current

    Row(
        modifier = modifier
            .clip(RoundedCornerShape(QuickInkRadius.pill))
            .background(if (active) colors.surface else Color.Transparent)
            .clickable(onClick = onClick)
            .padding(vertical = QuickInkSpacing.s2),
        verticalAlignment     = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Center,
    ) {
        Icon(
            imageVector       = icon,
            contentDescription = null,
            tint              = if (active) colors.ink else colors.inkSoft,
            modifier          = Modifier.size(13.dp),
        )
        Spacer(Modifier.size(6.dp))
        Text(
            text  = label,
            style = type.label,
            color = if (active) colors.ink else colors.inkSoft,
        )
    }
}

File: apps/quickink/ios/QuickInk/Scan/PageEditorScreen.swift
/*
 * PageEditorScreen.swift
 *
 * Page-editing surface — re-crop, rotate, retake. Reachable
 * from `ScanReviewScreen` after a capture completes (before the
 * user taps Done) and from `NoteEditorScreen`'s page tab when the
 * user wants to revise an existing scan.
 *
 * This is a UI wireframe today — the actual image transform
 * pipeline (Core Graphics rotation, draggable crop corners, page
 * re-capture via VisionKit) lands in a follow-up. Shipping the
 * picker chrome now means the route + theme + control surface
 * are settled when the transforms slot in.
 *
 * Mirror of Android `PageEditorScreen.kt`.
 */

import SwiftUI
#if canImport(UIKit)
import UIKit
#endif

struct PageEditorScreen: View {
    /// Page image (or page-preview URL) under edit. Today the
    /// mock renders a lined-paper placeholder when `imageURL` is
    /// nil; once the scan pipeline hands a real URL through, it
    /// renders the captured image.
    var imageURL: URL?

    let onCancel: () -> Void
    let onSave: () -> Void
    let onRetake: () -> Void

    @State private var rotation: Double = 0
    @State private var activeTool: Tool = .crop

    enum Tool { case crop, rotate, retake }

    var body: some View {
        VStack(spacing: 0) {
            topBar
            Spacer()
            pageCanvas
                .padding(QuickInkSpacing.s5)
            Spacer()
            toolStrip
                .padding(.horizontal, QuickInkSpacing.s5)
                .padding(.bottom, QuickInkSpacing.s4)
            actionRow
                .padding(.horizontal, QuickInkSpacing.s5)
                .padding(.bottom, QuickInkSpacing.s7)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(QuickInkColors.bg.ignoresSafeArea())
    }

    // MARK: - Top bar

    @ViewBuilder
    private var topBar: some View {
        HStack {
            Button(action: onCancel) {
                Text("Cancel")
                    .font(QuickInkText.label)
                    .foregroundStyle(QuickInkColors.inkSoft)
                    .padding(QuickInkSpacing.s3)
            }
            .buttonStyle(.plain)

            Spacer()

            Text("Edit page")
                .font(QuickInkText.label)
                .foregroundStyle(QuickInkColors.ink)

            Spacer()

            Button(action: onSave) {
                Text("Save")
                    .font(QuickInkText.label)
                    .foregroundStyle(QuickInkColors.accent)
                    .padding(QuickInkSpacing.s3)
            }
            .buttonStyle(.plain)
        }
        .padding(.horizontal, QuickInkSpacing.s2)
        .padding(.top, QuickInkSpacing.s2)
    }

    // MARK: - Page canvas

    @ViewBuilder
    private var pageCanvas: some View {
        ZStack {
            // Image area — real captured image when available,
            // lined-paper placeholder otherwise.
            Group {
                #if canImport(UIKit)
                if let url = imageURL, let image = loadImage(url: url) {
                    Image(uiImage: image)
                        .resizable()
                        .scaledToFit()
                } else {
                    placeholderPage
                }
                #else
                placeholderPage
                #endif
            }
            .rotationEffect(.degrees(rotation))
            .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous)
                    .stroke(QuickInkColors.border, lineWidth: 1)
            )

            // Crop handles overlay — only when crop tool active.
            if activeTool == .crop {
                cropHandlesOverlay
            }
        }
    }

    @ViewBuilder
    private var placeholderPage: some View {
        ZStack {
                        QuickInkLinedPaper(
                            tone: QuickInkColors.surface,
                            lineSpacing: 16,
                            lineOpacity: 0.10
                        )
                        HStack(spacing: 0) {
                            Rectangle()
                                .fill(QuickInkColors.accent.opacity(0.6))
                                .frame(width: 1.5)
                                .padding(.leading, 32)
                            Spacer()
                        }
                        .padding(.vertical, 16)
                        Text("scanned page")
                            .font(QuickInkFont.handwritten(28))
                            .foregroundStyle(QuickInkColors.inkSoft.opacity(0.7))
        }
    }

    @ViewBuilder
    private var cropHandlesOverlay: some View {
        GeometryReader { geo in
            let inset: CGFloat = 16
            ZStack {
                cropCorner.position(x: inset,                   y: inset)
                cropCorner.position(x: geo.size.width - inset,  y: inset)
                cropCorner.position(x: inset,                   y: geo.size.height - inset)
                cropCorner.position(x: geo.size.width - inset,  y: geo.size.height - inset)
            }
        }
    }

    @ViewBuilder
    private var cropCorner: some View {
        Circle()
            .fill(QuickInkColors.accent)
            .frame(width: 14, height: 14)
            .shadow(color: QuickInkColors.accent.opacity(0.4), radius: 6, x: 0, y: 0)
    }

    // MARK: - Tool strip (segmented)

    @ViewBuilder
    private var toolStrip: some View {
        HStack(spacing: 4) {
            tool(.crop,   icon: "crop",                   label: "Crop")
            tool(.rotate, icon: "rotate.right",           label: "Rotate")
            tool(.retake, icon: "arrow.triangle.2.circlepath", label: "Retake")
        }
        .padding(4)
        .background(QuickInkColors.borderSoft)
        .clipShape(RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous))
    }

    @ViewBuilder
    private func tool(_ kind: Tool, icon: String, label: String) -> some View {
        let active = (activeTool == kind)
        Button(action: { handleTool(kind) }) {
            HStack(spacing: 6) {
                Image(systemName: icon)
                    .font(.system(size: 13, weight: .medium))
                Text(label)
                    .font(QuickInkText.label)
            }
            .foregroundStyle(active ? QuickInkColors.ink : QuickInkColors.inkSoft)
            .frame(maxWidth: .infinity)
            .padding(.vertical, QuickInkSpacing.s2)
            .background(
                RoundedRectangle(cornerRadius: QuickInkRadius.pill, style: .continuous)
                    .fill(active ? QuickInkColors.surface : .clear)
                    .shadow(color: active ? QuickInkColors.ink.opacity(0.06) : .clear, radius: 4, x: 0, y: 2)
            )
        }
        .buttonStyle(.plain)
    }

    private func handleTool(_ kind: Tool) {
        activeTool = kind
        switch kind {
        case .crop:
            // Already shows handles — interaction is the follow-up.
            break
        case .rotate:
            // Step 90° per tap. Real rotation also flips the
            // crop rect when implemented.
            withAnimation(.easeInOut(duration: 0.18)) {
                rotation += 90
                if rotation >= 360 { rotation = 0 }
            }
        case .retake:
            onRetake()
        }
    }

    // MARK: - Action row (placeholder hint)

    @ViewBuilder
    private var actionRow: some View {
        HStack(spacing: QuickInkSpacing.s2) {
            Image(systemName: "info.circle")
                .font(.system(size: 12))
                .foregroundStyle(QuickInkColors.muted)
            Text("Drag the corners to crop · ⟲ rotates 90°")
                .font(QuickInkText.caption)
                .foregroundStyle(QuickInkColors.muted)
        }
    }

    // MARK: - Image loader

    private func loadImage(url: URL) -> UIImage? {
        // Loaded synchronously — fine for a single small page
        // preview. When we swap in the real scan pipeline (large
        // PDFs / multi-page), move this to an async load with a
        // placeholder while in flight.
        guard let data = try? Data(contentsOf: url) else { return nil }
        return UIImage(data: data)
    }
}
